src-electron/generator/helper-endpointconfig.js

```diff
@@ -465,7 +465,10 @@ function endpoint_attribute_long_defaults(options) {
 }
 
 function asMEI(manufacturerCode, code) {
-  return '0x' + bin.int32ToHex((manufacturerCode << 16) + code)
+  // Left-shift (and for that matter bitwise or) produces a _signed_ 32-bit
+  // number, which will probably be negative.  Force it to unsigned 32-bit using
+  // >>> 0.
+  return '0x' + bin.int32ToHex(((manufacturerCode << 16) | code) >>> 0);
```

Comment:
Shouldn't the >>>0 be applied to manufacturerCode (or the << 16 expression), or is it OK to do after the full expression (i.e. start interpreting 2's complement value as unsigned)? Not super familiar with that corner of Javascript, but this is a bit odd.

Perhaps a less tricky alternative is:

```
  return '0x' + bin.int32ToHex(manufacturerCode) + bin.int32ToHex(code).padStart(4, '0');
```

Functions like asHex(val, numBitsWidth) and asZeroPaddedHex(val, numBitsWidth) would probably hex with composing more of these if needed.

Reply:
Both the `<<` operator and the `|` operator produce signed output in JavaScript.  See https://262.ecma-international.org/6.0/#sec-left-shift-operator-runtime-semantics-evaluation step 12 and https://262.ecma-international.org/6.0/#sec-binary-bitwise-operators-runtime-semantics-evaluation steps 7 and 11.

So `(a << b)>>>0 | c` would shift by `b`, then return an unsigned integer, which would get converted to a signed integer, and the bitwise or would happen on signed integers....

So the `>>>` really does need to be on the final result.  I totally agree the bitwise ops in JS are _very_ odd.

> Perhaps a less tricky alternative is:

This works only if `code` does not already included the `manufacturerCode` in its high bits.  Which is a property the old approach had as well, and might be OK, since including the manufacturer code in the high bits of `code` if you also set `manufacturerCode` is arguably pilot error...

So maybe I should just not worry about handling that case "gracefully" and just fix the actual correctness bug around signed values.

Reply:
>   `return '0x' + bin.int32ToHex(manufacturerCode) + bin.int32ToHex(code).padStart(4, '0');`

No, this does not do the right thing at all: it will create a string of length 2 + 8 + 8 = 18, whereas we want a string of length at most 10 here.

Since code might in fact be a full 32-bit value, we _have_ to combine `manufacturerCode` with `code` as integers, not as strings.